src/main/kotlin/cacophony/automata/DFAJoin.kt

```diff
@@ -0,0 +1,83 @@
+package cacophony.automata
+
+import cacophony.automata.minimalization.via
+
+// Joins given list of the automata into a single DFA with accepting states marked
+// with a Result corresponding to the source DFAs.
+// Throws if any state would be marked by many results, i.e. when any two DFAs accept the same word.
+fun <State, Atom, Result> joinAutomata(
+    automataAndResults: List<Pair<GenericDFA<State, Atom, *>, Result>>,
+): GenericDFA<Int, Atom, Result?> {
```

Comment:
with @Giselus we decided that `Result` param in DFA will be non-nullable, and instead we recognize that state is not accepting iff it does not exist as key of the `results` map

Reply:
That was my first version also, changed when I saw @Giselus PR. Will change it back